lib/auth/export.go

```diff
@@ -0,0 +1,228 @@
+	ExportAuthorityFingerprint string
+	ExportPrivateKeys          bool
+	CompatVersion              string
+}
+
+// ExportAuthorities returns the list of authorities in OpenSSH compatible formats
+// If the ExportAuthoritiesRequest.AuthType is present only prints keys for CAs of this type,
+// otherwise returns all keys concatenated
+func ExportAuthorities(ctx context.Context, req ExportAuthoritiesRequest) (string, error) {
+	ret := strings.Builder{}
+
+	var typesToExport []types.CertAuthType
+
+	// this means to export TLS authority
+	switch req.AuthType {
+	// "tls" is supported for backwards compatibility.
+	// "tls-host" and "tls-user" were added later to allow export of the user
+	// TLS CA.
+	case "tls", "tls-host":
+		return exportTLSAuthority(ctx, req.Client, types.HostCA, false, req.ExportPrivateKeys)
+	case "tls-user":
+		return exportTLSAuthority(ctx, req.Client, types.UserCA, false, req.ExportPrivateKeys)
+	case "db":
+		return exportTLSAuthority(ctx, req.Client, types.DatabaseCA, false, req.ExportPrivateKeys)
+	case "tls-user-der", "windows":
+		return exportTLSAuthority(ctx, req.Client, types.UserCA, true, req.ExportPrivateKeys)
+	}
+
+	// if no --type flag is given, export HostCA and UserCA.
+	if req.AuthType == "" {
+		typesToExport = []types.CertAuthType{types.HostCA, types.UserCA}
+	} else {
+		authType := types.CertAuthType(req.AuthType)
+		if err := authType.Check(); err != nil {
+			return "", trace.Wrap(err)
+		}
+		typesToExport = []types.CertAuthType{authType}
+	}
+	localAuthName, err := req.Client.GetDomainName(ctx)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	// fetch authorities via auth API (and only take local CAs, ignoring
+	// trusted ones)
+	var authorities []types.CertAuthority
+	for _, at := range typesToExport {
+		cas, err := req.Client.GetCertAuthorities(ctx, at, req.ExportPrivateKeys)
+		if err != nil {
+			return "", trace.Wrap(err)
+		}
+		for _, ca := range cas {
+			if ca.GetClusterName() == localAuthName {
+				authorities = append(authorities, ca)
+			}
+		}
+	}
+
+	for _, ca := range authorities {
+		if req.ExportPrivateKeys {
+			for _, key := range ca.GetActiveKeys().SSH {
+				fingerprint, err := sshutils.PrivateKeyFingerprint(key.PrivateKey)
+				if err != nil {
+					return "", trace.Wrap(err)
+				}
+				if req.ExportAuthorityFingerprint != "" && fingerprint != req.ExportAuthorityFingerprint {
+					continue
+				}
+				ret.WriteString(string(key.PrivateKey) + "\n")
+			}
+		} else {
+			for _, key := range ca.GetTrustedSSHKeyPairs() {
+				fingerprint, err := sshutils.AuthorizedKeyFingerprint(key.PublicKey)
+				if err != nil {
+					return "", trace.Wrap(err)
+				}
+				if req.ExportAuthorityFingerprint != "" && fingerprint != req.ExportAuthorityFingerprint {
+					continue
+				}
+
+				// export certificates in the old 1.0 format where host and user
+				// certificate authorities were exported in the known_hosts format.
+				if req.CompatVersion == "1.0" {
```

Comment:
What about other compat versions?

Should we make CompatVersion an enum?

Reply:
I would say we don't need to support other compat versions.
I'm not sure what those would be, but adding features to the export functionality is not necessary at this point.

Given the above, converting into an enum seems unnecessary.

Reply:
I think I would still take an enum over a string, that makes it clear that we only really support "no compat" and "v1".

Reply:
Changing this to a boolean as you suggested below